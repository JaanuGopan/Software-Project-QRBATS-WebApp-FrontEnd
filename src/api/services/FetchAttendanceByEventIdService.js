// FetchEventsService.js
import axios from "axios";
import ApiConstants from "../config/ApiConstants";

class FetchAttendanceByEventIdService {
  static async fetchAttendance(eventId) {
    try {
      const response = await axios.post(
        ApiConstants.getAttendanceByEventIdUrl,
        {
          eventId: eventId,
        }
      );

      return response.data;
    } catch (error) {
      console.error("Error fetching attendance:", error);
      throw error; // Re-throw the error to handle it in the component
    }
  }
}

export default FetchAttendanceByEventIdService;
