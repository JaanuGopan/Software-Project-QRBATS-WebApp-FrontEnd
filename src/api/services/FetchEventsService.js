// FetchEventsService.js
import axios from "axios";
import ApiConstants from "../config/ApiConstants";

class FetchEventsService {
  static async fetchEvents() {
    try {
      const response = await axios.post(ApiConstants.getAllEventUrl);
      return response.data;
    } catch (error) {
      console.error("Error fetching events:", error);
      throw error; // Re-throw the error to handle it in the component
    }
  }
}

export default FetchEventsService;
