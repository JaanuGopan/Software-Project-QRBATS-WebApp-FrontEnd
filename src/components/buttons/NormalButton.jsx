import React from 'react';
import './normal-button.css';

const NormalButton = ({ title, icon, handleClick }) => {
  return (
    <div className="row my-1 align-items-center normal-button-container" onClick={handleClick}>
      <div className="col">
        <div className="row py-2">
          <div className="col-auto pe-0">{icon}</div>
          <div className="col-auto fs-6">{title}</div>
        </div>
      </div>
    </div>
  );
};

export default NormalButton;
