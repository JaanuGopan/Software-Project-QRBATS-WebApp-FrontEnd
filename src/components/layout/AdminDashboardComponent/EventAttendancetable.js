import React, { useState, useEffect } from "react";
import "../../../pages/AdminDashboard/AdminDashboard.css";
import html2canvas from "html2canvas";
import jsPDF from "jspdf";

const EventAttendancetable = ({
  search,
  attendanceList,
  handleGeneratePDF,
}) => {
  const [attendance, setAttendance] = useState([]);
  const [selectedAttendance, setSelectedAttendance] = useState(null);

  useEffect(() => {
    // Update events whenever eventList prop changes
    setAttendance(attendanceList);
  }, [attendanceList]);

  const handleAttendanceClick = (attendance) => {
    setSelectedAttendance(attendance);
  };

  return (
    <div className="tableDesign">
      <table className="student-tableArrangement">
        <thead>
          <tr>
            <th>No</th>
            <th className="expand">Student Name</th>
            <th>Registration Number</th>
            <th>Attendance Date</th>
            <th>Attendance Time</th>
          </tr>
        </thead>

        <tbody>
          {attendance
            .filter(
              (attendance) =>
                attendance.studentName
                  .toLowerCase()
                  .includes(search.toLowerCase()) ||
                attendance.indexNumber
                  .toLowerCase()
                  .includes(search.toLowerCase())
            )
            .map((attendance, index) => (
              <tr
                key={index}
                className={
                  selectedAttendance === attendance
                    ? "selected-row"
                    : "event-row"
                }
              >
                <td>{index + 1}</td>
                <td>{attendance.studentName}</td>
                <td>{attendance.indexNumber}</td>
                <td>{attendance.attendedDate}</td>
                <td>{attendance.attendedTime}</td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};

export default EventAttendancetable;
