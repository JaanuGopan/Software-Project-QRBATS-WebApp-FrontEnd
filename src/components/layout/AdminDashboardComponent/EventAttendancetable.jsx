import React, { useState, useEffect } from 'react';
import '../../../pages/AdminDashboard/AdminDashboard.css';
import html2canvas from 'html2canvas';
import jsPDF from 'jspdf';
import { FaCheck } from 'react-icons/fa';
import { AiOutlineClose } from 'react-icons/ai';

const EventAttendanceTable = ({ search, attendanceList, handleGeneratePDF }) => {
  const [attendance, setAttendance] = useState([]);
  const [selectedAttendance, setSelectedAttendance] = useState(null);

  useEffect(() => {
    // Update events whenever eventList prop changes
    setAttendance(attendanceList);
  }, [attendanceList]);

  const handleAttendanceClick = (attendance) => {
    setSelectedAttendance(attendance);
  };

  return (
    <div className="tableDesign">
      <table className="student-tableArrangement">
        <thead>
          <tr>
            <th>No</th>
            <th className="expand">Student Name</th>
            <th>Registration Number</th>
            <th>Attendance Date</th>
            <th>Attendance Time</th>
            <th>Attendance Status</th>
          </tr>
        </thead>

        {attendance.filter(
          (attendance) =>
            attendance.studentName.toLowerCase().includes(search.toLowerCase()) ||
            attendance.indexNumber.toLowerCase().includes(search.toLowerCase())
        ).length > 0 ? (
          <tbody>
            {attendance
              .filter(
                (attendance) =>
                  attendance.studentName.toLowerCase().includes(search.toLowerCase()) ||
                  attendance.indexNumber.toLowerCase().includes(search.toLowerCase())
              )
              .map((attendance, index) => (
                <tr
                  key={index}
                  className={selectedAttendance === attendance ? 'selected-row' : 'event-row'}
                >
                  <td>{index + 1}</td>
                  <td>{attendance.studentName}</td>
                  <td>{attendance.indexNumber}</td>
                  <td>{attendance.attendedDate}</td>
                  <td>{attendance.attendedTime}</td>
                  <td>
                    {attendance.attendanceStatus ? (
                      <FaCheck color="green" />
                    ) : (
                      <AiOutlineClose color="red" />
                    )}
                  </td>
                </tr>
              ))}
          </tbody>
        ) : (
          <tbody className="nodata">No Data Available</tbody>
        )}
      </table>
    </div>
  );
};

export default EventAttendanceTable;
