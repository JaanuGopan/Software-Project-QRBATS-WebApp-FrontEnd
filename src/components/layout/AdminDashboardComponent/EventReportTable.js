import React, { useState, useEffect } from "react";
import "../../../pages/AdminDashboard/AdminDashboard.css";
import { CiViewList } from "react-icons/ci";
import axios from "axios";
import FetchAttendanceByEventIdService from "../../../api/services/FetchAttendanceByEventIdService";

const EventReportTable = ({
  handleOpenReportWindow,
  search,
  onEventClick,
  eventList,
  attendedStudentList,
}) => {
  const [events, setEvents] = useState([]);
  const [selectedEvent, setSelectedEvent] = useState(null);
  const [attendanceList, setAttendanceList] = useState([]);

  useEffect(() => {
    // Update events whenever eventList prop changes
    setEvents(eventList);
  }, [eventList]);

  const handleEventClick = (event) => {
    setSelectedEvent(event);
    onEventClick(event);
  };

  const handleAttendanceList = (attendanceStudentList) => {
    setAttendanceList(attendanceStudentList);
    attendedStudentList(attendanceStudentList);
  };

  const handleReloadAttendanceList = async (eventId) => {
    FetchAttendanceByEventIdService.fetchAttendance(eventId)
      .then((attendanceList) => {
        setAttendanceList(attendanceList);
        handleAttendanceList(attendanceList);
        handleOpenReportWindow();
      })
      .catch((error) => {
        console.error("Error fetching attendance:", error);
      });
  };

  return (
    <div className="tableDesign">
      <table className="event-report-tableArrangement">
        <thead>
          <tr>
            <th>No</th>
            <th className="expand">Name</th>
            <th>Start Date</th>
            <th>End Date</th>
            <th>Venue</th>
            <th>Start Time</th>
            <th>End Time</th>
            <th>View</th>
          </tr>
        </thead>
        <tbody>
          {events
            .filter(
              (event) =>
                event.eventName.toLowerCase().includes(search.toLowerCase()) ||
                event.eventVenue.toLowerCase().includes(search.toLowerCase())
            )
            .map((event, index) => (
              <tr
                key={index}
                onClick={() => handleEventClick(event)}
                className={
                  selectedEvent === event ? "selected-row" : "event-row"
                }
              >
                <td>{index + 1}</td>
                <td>{event.eventName}</td>
                <td>{event.eventDate}</td>
                <td>{event.eventValidDate}</td>
                <td>{event.eventVenue}</td>
                <td>{event.eventTime}</td>
                <td>{event.eventEndTime}</td>
                <td>
                  <button
                    onClick={() => handleReloadAttendanceList(event.eventId)}
                    className="ViewButton"
                  >
                    <CiViewList className="EditIcon" />
                    <p className="ViewButtonLabel">View</p>
                  </button>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};

export default EventReportTable;
