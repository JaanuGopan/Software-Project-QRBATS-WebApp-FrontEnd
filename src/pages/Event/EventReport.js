import React, { useState, useEffect } from "react";
import "../Staff/StaffA.css";
import EventReportTable from "../../components/layout/AdminDashboardComponent/EventReportTable";
import EventAttendancetable from "../../components/layout/AdminDashboardComponent/EventAttendancetable";
import NormalButton from "../../components/layout/AdminDashboardComponent/NormalButton";
import { MdArrowBack } from "react-icons/md";
import { BiSolidPrinter } from "react-icons/bi";
import FetchEventsService from "../../api/services/FetchEventsService";
import FetchAttendanceByEventIdService from "../../api/services/FetchAttendanceByEventIdService";
import html2canvas from "html2canvas";
import jsPDF from "jspdf";

const EventReport = () => {
  const [eventReportTable, setEventReportTable] = useState(true);

  const [selectedEvent, setSelectedEvent] = useState(null);
  const [eventList, setEventList] = useState([]);
  const [search, setSearch] = useState("");

  const [selectedAttendance, setSelectedAttendance] = useState(null);
  const [attendanceList, setAttendanceList] = useState([]);

  useEffect(() => {
    // Fetch the list of events from the API using the new class
    handleReloadEventList();
  }, []);

  const handleEventClick = (event) => {
    setSelectedEvent(event);
    // Do whatever you want with the selected event data
    console.log("Selected Event:", event);
  };

  const handleReloadEventList = async () => {
    FetchEventsService.fetchEvents()
      .then((events) => {
        setEventList(events);
      })
      .catch((error) => {
        console.error("Error fetching events:", error);
      });
  };

  const handleAttendenceList = (attendanceStudentList) => {
    setAttendanceList(attendanceStudentList);
  };

  const generatePDF = () => {
    const input = document.getElementById("table-to-print");
    html2canvas(input).then((canvas) => {
      const imgData = canvas.toDataURL("image/png");
      const pdf = new jsPDF();
      pdf.addImage(imgData, "PNG", 0, 0);
      pdf.save("table.pdf");
    });
  };

  const handleGeneratePDF = () => {
    generatePDF();
  };

  return (
    <div className="staff-Dash">
      {eventReportTable ? (
        <div>
          <div className="staff-SearchEvent">
            <p className="staff-mainHead">All Event Details</p>
            <input
              type="text"
              placeholder="Search..."
              style={{
                width: "150px",
                padding: "3px 40px",
                border: "0.5px solid black",
                borderRadius: "5px",
                textAlign: "center",
              }}
              onChange={(e) => setSearch(e.target.value)}
            />
          </div>
          <div className="staff-EventList">
            <EventReportTable
              handleOpenReportWindow={() => setEventReportTable(false)}
              search={search}
              onEventClick={handleEventClick}
              eventList={eventList}
              attendedStudentList={handleAttendenceList}
            />
          </div>
        </div>
      ) : (
        <div>
          <div className="staff-SearchEvent">
            <p className="staff-mainHead">Event Attendance Details</p>
            <input
              type="text"
              placeholder="Search..."
              style={{
                width: "150px",
                padding: "3px 40px",
                border: "0.5px solid black",
                borderRadius: "5px",
                textAlign: "center",
              }}
              onChange={(e) => setSearch(e.target.value)}
            />
          </div>

          <div id="table-to-print" className="staff-EventList">
            <EventAttendancetable
              search={search}
              attendanceList={attendanceList}
            />
            <div className="staff-List-Buttons">
              <NormalButton
                handleClick={() => setEventReportTable(true)}
                title={"Back"}
                titlewithiconicon={<MdArrowBack className="staff-buttonIcon" />}
              />
              <NormalButton
                title={"Print"}
                handleClick={handleGeneratePDF}
                titlewithiconicon={
                  <BiSolidPrinter className="staff-buttonIcon" />
                }
              />
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default EventReport;
