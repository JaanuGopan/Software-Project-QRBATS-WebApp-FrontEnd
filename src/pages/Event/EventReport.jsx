import React, { useState, useEffect, use, useContext } from 'react';
import '../Staff/StaffA.css';
import EventReportTable from '../../components/layout/AdminDashboardComponent/EventReportTable';
import EventAttendanceTable from '../../components/layout/AdminDashboardComponent/EventAttendancetable';
import NormalButton from '../../components/layout/AdminDashboardComponent/NormalButton';
import { MdArrowBack } from 'react-icons/md';
import { BiSolidPrinter } from 'react-icons/bi';
import html2canvas from 'html2canvas';
import jsPDF from 'jspdf';
import EventService from '../../api/services/EventService';
import AttendanceService from '../../api/services/AttendanceService';
import { toast } from 'react-toastify';
import { CircularProgress } from '@mui/material';
import { AuthContext } from '../../config/AuthProvider';

const EventReport = () => {
  const [eventReportTable, setEventReportTable] = useState(true);
  const [selectedEvent, setSelectedEvent] = useState(null);
  const [eventList, setEventList] = useState([]);
  const [search, setSearch] = useState('');
  const [selectedAttendance, setSelectedAttendance] = useState(null);
  const [attendanceList, setAttendanceList] = useState([]);
  const { user } = useContext(AuthContext);
  const { userId } = user || {};

  useEffect(() => {
    handleReloadEventList();
  }, []);

  const handleEventClick = (event) => {
    setSelectedEvent(event);
    // Do whatever you want with the selected event data
    console.log('Selected Event:', event);
  };

  const handleReloadEventList = async () => {
    EventService.getEventByUserID(userId)
      .then((events) => {
        setEventList(events);
      })
      .catch((error) => {
        console.error('Error fetching events:', error);
      });
  };

  const handleAttendanceList = (attendanceStudentList) => {
    setAttendanceList(attendanceStudentList);
  };

  const generatePDF = () => {
    const input = document.getElementById('table-to-print');
    html2canvas(input).then((canvas) => {
      const imgData = canvas.toDataURL('image/png');
      const pdf = new jsPDF();
      pdf.addImage(imgData, 'PNG', 0, 0);
      pdf.save('table.pdf');
    });
  };

  const [loadingDownloadReport, setLoadingDownloadReport] = useState(false);

  const handleDownloadAttendanceEventReport = async (eventId) => {
    try {
      setLoadingDownloadReport(true);
      const response = await AttendanceService.downloadEventAttendance(eventId);
      if (response.status === 200) {
        const data = response.data;
        const blob = new Blob([data], { type: 'text/csv' });
        const link = document.createElement('a');
        link.href = URL.createObjectURL(blob);
        link.download = `event_${eventId}_attendance.csv`;
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
      } else if (response.status === 400) {
        toast.error(response.data);
      } else {
        toast.error('Error In Downloading Report.');
      }
    } finally {
      setLoadingDownloadReport(false);
    }
  };

  return (
    <div className="staff-Dash">
      {eventReportTable ? (
        <div>
          <div className="staff-SearchEvent">
            <p className="staff-mainHead">All Event Details</p>
            <input
              type="text"
              placeholder="Search..."
              style={{
                width: '150px',
                padding: '3px 40px',
                border: '0.5px solid black',
                borderRadius: '5px',
                textAlign: 'center',
              }}
              onChange={(e) => setSearch(e.target.value)}
            />
          </div>
          <div className="staff-EventList">
            <EventReportTable
              handleOpenReportWindow={() => setEventReportTable(false)}
              search={search}
              onEventClick={handleEventClick}
              eventList={eventList}
              attendedStudentList={handleAttendanceList}
            />
          </div>
        </div>
      ) : (
        <div>
          <div className="staff-SearchEvent">
            <p className="staff-mainHead">Event Attendance Details</p>
            <input
              type="text"
              placeholder="Search..."
              style={{
                width: '150px',
                padding: '3px 40px',
                border: '0.5px solid black',
                borderRadius: '5px',
                textAlign: 'center',
              }}
              onChange={(e) => setSearch(e.target.value)}
            />
          </div>

          <div id="table-to-print" className="staff-EventList">
            <EventAttendanceTable search={search} attendanceList={attendanceList} />
            <div className="staff-List-Buttons">
              <NormalButton
                handleClick={() => setEventReportTable(true)}
                title={'Back'}
                titlewithiconicon={<MdArrowBack className="staff-buttonIcon" />}
              />
              {loadingDownloadReport ? (
                <CircularProgress />
              ) : (
                <NormalButton
                  title={'Print'}
                  handleClick={() => handleDownloadAttendanceEventReport(selectedEvent.eventId)}
                  titlewithiconicon={<BiSolidPrinter className="staff-buttonIcon" />}
                />
              )}
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default EventReport;
