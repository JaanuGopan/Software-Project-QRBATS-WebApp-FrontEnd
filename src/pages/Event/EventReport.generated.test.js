import renderer from 'react-test-renderer';
import React, { useState, useEffect } from "react";

import EventReportTable from "../../components/layout/AdminDashboardComponent/EventReportTable";
import EventAttendancetable from "../../components/layout/AdminDashboardComponent/EventAttendancetable";
import NormalButton from "../../components/layout/AdminDashboardComponent/NormalButton";
import { MdArrowBack } from "react-icons/md";
import { BiSolidPrinter } from "react-icons/bi";
import FetchEventsService from "../../api/services/FetchEventsService";
import FetchAttendanceByEventIdService from "../../api/services/FetchAttendanceByEventIdService";
import html2canvas from "html2canvas";
import jsPDF from "jspdf";
import EventReport from "./EventReport";

jest.mock("../Staff/StaffA.css");
jest.mock("../../components/layout/AdminDashboardComponent/EventReportTable");
jest.mock("../../components/layout/AdminDashboardComponent/EventAttendancetable");
jest.mock("../../components/layout/AdminDashboardComponent/NormalButton");
jest.mock("react-icons/md");
jest.mock("react-icons/bi");
jest.mock("../../api/services/FetchEventsService");
jest.mock("../../api/services/FetchAttendanceByEventIdService");
jest.mock("html2canvas");
jest.mock("jspdf");

const renderTree = tree => renderer.create(tree);
describe('<EventReport>', () => {
  it('should render component', () => {
    expect(renderTree(<EventReport 
    />).toJSON()).toMatchSnapshot();
  });
  
});