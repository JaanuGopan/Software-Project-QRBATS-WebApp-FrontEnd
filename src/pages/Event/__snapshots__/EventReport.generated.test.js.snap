// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventReport> should render component 1`] = `
<div
  className="staff-Dash"
>
  <div>
    <div
      className="staff-SearchEvent"
    >
      <p
        className="staff-mainHead"
      >
        All Event Details
      </p>
      <input
        onChange={[Function]}
        placeholder="Search..."
        style={
          Object {
            "border": "0.5px solid black",
            "borderRadius": "5px",
            "padding": "3px 40px",
            "textAlign": "center",
            "width": "150px",
          }
        }
        type="text"
      />
    </div>
    <div
      className="staff-EventList"
    />
  </div>
</div>
`;
